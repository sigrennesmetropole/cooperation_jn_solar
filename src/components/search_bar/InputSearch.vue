<script setup lang="ts">
import { ref } from 'vue'
import iconMagnifyingGlass from '@/assets/icons/magnifying-glass.svg'
import iconMultiply from '@/assets/icons/icon-multiply.svg'
import iconFilterSearchBar from '@/assets/icons/icon-filter-searchbar.svg'
import { useAddressStore } from '@/stores/address'

defineProps({
  isDisplayAutocompletion: {
    type: Boolean,
    default: false,
  },
  isEmptySearch: {
    type: Boolean,
    default: false,
  },
  sizeBeginSearch: {
    type: Number,
  },
})

const search = ref('')
const emit = defineEmits(['emptySearch', 'toggleFilters', 'inputKeyUp'])
const addressStore = useAddressStore()

function emptySearch() {
  search.value = ''
  emit('emptySearch')
}

addressStore.$subscribe(() => {
  search.value = addressStore.address
})
</script>

<template>
  <div
    class="flex flex-row items-center h-11 shadow-lg rounded-lg p-0 mb-0 bg-white w-[402px]"
    :class="
      isDisplayAutocompletion || isEmptySearch ? 'border border-black' : ''
    "
  >
    <div class="flex flex-row items-center justify-center w-8 h-full ml-2">
      <img :src="iconMagnifyingGlass" class="w-4 h-4" />
    </div>
    <input
      id="autocomplete-input"
      class="w-full h-full border-none text-base font-medium font-dm-sans placeholder-black focus:ring-0 !border-black"
      type="text"
      placeholder="Entrez votre adresse ici"
      v-model="addressStore.address"
      @keyup="$emit('inputKeyUp', search)"
    />
    <button
      class="flex flex-row items-center justify-center w-11 h-full cursor-pointer"
      @click="emptySearch()"
      v-if="search !== ''"
      id="button-empty-search"
    >
      <img :src="iconMultiply" class="w-4 h-4" />
    </button>
    <button
      class="flex flex-row items-center justify-center w-11 h-full cursor-pointer border-l border-neutral-200"
      @click="$emit('toggleFilters')"
      id="toggle-filters"
    >
      <img :src="iconFilterSearchBar" class="w-4 h-4" />
    </button>
  </div>
</template>
