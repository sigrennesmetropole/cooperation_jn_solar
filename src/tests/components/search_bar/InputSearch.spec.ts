import InputSearch from '@/components/search_bar/InputSearch.vue'
import { describe, it, expect, beforeEach, vi } from 'vitest'
import { mount, VueWrapper } from '@vue/test-utils'
import { createTestingPinia } from '@pinia/testing'

describe('InputSearch.vue', () => {
  let wrapper: VueWrapper

  beforeEach(async () => {
    wrapper = mount(InputSearch, {
      global: {
        plugins: [
          createTestingPinia({
            createSpy: vi.fn,
            stubActions: false,
            stubPatch: false,
            fakeApp: true,
          }),
        ],
      },
      propsData: {
        isDisplayAutocompletion: true,
        isEmptySearch: false,
        sizeBeginSearch: 0,
      },
    })
  })

  it('renders the component correctly', () => {
    // Check if the input element is rendered
    const inputElement = wrapper.find('input')
    expect(inputElement.exists()).toBe(true)

    // Check if the magnifying glass icon is rendered
    const magnifyingGlassIcon = wrapper.find('img[src$="magnifying-glass.svg"]')
    expect(magnifyingGlassIcon.exists()).toBe(true)

    // Check if the multiply icon is not rendered initially
    const multiplyIcon = wrapper.find('img[src$="icon-multiply.svg"]')
    expect(multiplyIcon.exists()).toBe(false)

    // Check if the filter icon is rendered
    const filterIcon = wrapper.find('img[src$="icon-filter-searchbar.svg"]')
    expect(filterIcon.exists()).toBe(true)
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('emits toggleFilters event when filter button is clicked', async () => {
    // Find the filter button and trigger a click event
    const filterButton = wrapper.find('button[id="toggle-filters"]')
    await filterButton.trigger('click')

    // Check if the toggleFilters event has been emitted
    expect(wrapper.emitted().toggleFilters).toBeTruthy()
  })

  it('emits inputKeyUp event with the search value when input is typed', async () => {
    const wrapper = mount(InputSearch)

    // Find the input element and set its value
    const inputElement = wrapper.find('input')
    await inputElement.setValue('Some address')

    // Trigger the keyup event
    await inputElement.trigger('keyup')

    // Check if the inputKeyUp event has been emitted with the search value
    expect(wrapper.emitted().inputKeyUp).toBeTruthy()
    expect(wrapper.emitted().inputKeyUp[0]).toEqual(['Some address'])
  })

  it('calls emptySearch function when the empty search button is clicked', async () => {
    const inputElement = wrapper.find('input')
    await inputElement.setValue('Some address')

    // Find the filter button and trigger a click event
    const filterButton = wrapper.find('button[id="button-empty-search"]')
    await filterButton.trigger('click')

    // Check if the toggleFilters event has been emitted
    expect(wrapper.emitted().emptySearch).toBeTruthy()
  })
})
