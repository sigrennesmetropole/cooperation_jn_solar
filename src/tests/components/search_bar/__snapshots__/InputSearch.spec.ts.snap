// Vitest Snapshot v1

exports[`InputSearch.vue > renders correctly 1`] = `
"<div class=\\"flex flex-row items-center h-11 shadow-lg rounded-lg p-0 mb-0 bg-white w-[402px] border border-black\\">
  <div class=\\"flex flex-row items-center justify-center w-8 h-full ml-2\\"><img src=\\"/src/assets/icons/magnifying-glass.svg\\" class=\\"w-4 h-4\\"></div><input id=\\"autocomplete-input\\" class=\\"w-full h-full border-none text-base font-medium font-dm-sans placeholder-black focus:ring-0 !border-black\\" type=\\"text\\" placeholder=\\"Entrez votre adresse ici\\">
  <!--v-if--><button class=\\"flex flex-row items-center justify-center w-11 h-full cursor-pointer border-l border-neutral-200\\" id=\\"toggle-filters\\"><img src=\\"/src/assets/icons/icon-filter-searchbar.svg\\" class=\\"w-4 h-4\\"></button>
</div>"
`;
